dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java

```diff
@@ -571,34 +578,83 @@ else if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.CHILDREN
     /**
      * Generates a sub query which provides a filter by organisation descendant
      * level.
+     *
+     * @param params the {@link EventQueryParams}
+     * @return
      */
-    private String getOrgUnitDescendantsClause( OrgUnitField orgUnitField,
-        List<DimensionalItemObject> dimensionOrFilterItems )
+    private String getOrgUnitDescendantsClause( EventQueryParams params )
     {
-        Map<String, List<OrganisationUnit>> collect = dimensionOrFilterItems.stream()
+        if ( params.getDimension( ORGUNIT_DIM_ID ) == null )
+        {
+            return StringUtils.EMPTY;
+        }
+
+        List<DimensionalItemObject> dimensionalItemObjects = params.getDimensionOrFilterItems( ORGUNIT_DIM_ID );
+
+        Map<Integer, List<OrganisationUnit>> collect = dimensionalItemObjects
+            .stream()
             .map( object -> (OrganisationUnit) object )
-            .collect( Collectors.groupingBy(
-                unit -> orgUnitField.getOrgUnitLevelCol( unit.getLevel(), getAnalyticsType() ) ) );
+            .collect( Collectors.groupingBy( OrganisationUnit::getLevel ) );
 
         return collect.keySet()
             .stream()
-            .map( org -> toInCondition( org, collect.get( org ) ) )
+            .map( org -> toInOrgUnitsCondition( org, collect.get( org ) ) )
             .collect( joining( " and " ) );
     }
 
     /**
      * Generates an in condition.
      *
-     * @param orgUnit the org unit identifier.
-     * @param organisationUnits the list of {@link OrganisationUnit}.
-     * @return a SQL in condition.
+     * @param level organisation unit level
+     * @param organisationUnits collection of user granted organisation units
+     * @return
      */
-    private String toInCondition( String orgUnit, List<OrganisationUnit> organisationUnits )
+    private String toInOrgUnitsCondition( int level, List<OrganisationUnit> organisationUnits )
     {
-        return organisationUnits.stream()
-            .filter( unit -> isNotEmpty( unit.getUid() ) )
-            .map( unit -> "'" + unit.getUid() + "'" )
-            .collect( joining( ",", orgUnit + OPEN_IN, ") " ) );
+        if ( organisationUnits == null || organisationUnits.isEmpty() )
+        {
+            return StringUtils.EMPTY;
+        }
+
+        List<String> allOrgUnitUidInLevel = getAllOrUnitIdsInLevel( level );
+
+        //is amount of requested level org units less than 50% of all level org units
+        if ( organisationUnits.size() == allOrgUnitUidInLevel.size()
```

Comment:
Some of this logic could maybe be factored out to a separate method to encapsulate and document the logic.